File: skills/ascii-art/scripts/img2ascii.py
#!/usr/bin/env python3
"""img2ascii.py — convert an image to ASCII/ANSI art in the terminal.

Uses, in order of preference: `chafa`, `jp2a`, `viu`, `img2txt` (libcaca),
then Pillow. Falls back gracefully with an install hint.

Examples:
  python3 img2ascii.py logo.png
  python3 img2ascii.py photo.jpg --width 80
  python3 img2ascii.py photo.jpg --color
"""

import argparse
import shutil
import subprocess
import sys

RAMP = " .,:;irsXA253hMHGS#9B&@"


def try_cli(path, args):
    commands = [
        (["chafa", "--symbols", "block+border+space", "--format", "symbols"], {}),
        (["jp2a", "--width"], {"width": True}),
        (["viu", "-w"], {"width": True}),
        (["img2txt", "-W"], {"width": True}),
    ]
    for cmd, opts in commands:
        exe = shutil.which(cmd[0])
        if not exe:
            continue
        argv = list(cmd)
        if opts.get("width") and args.width:
            argv += [str(args.width)]
        if cmd[0] == "chafa" and args.width:
            argv += ["--size", f"{args.width}x"]
        if cmd[0] == "chafa" and args.color:
            argv += ["--colors", "full"]
        argv.append(path)
        try:
            out = subprocess.run(argv, capture_output=True, text=True, timeout=30)
            if out.returncode == 0 and out.stdout.strip():
                print(out.stdout.rstrip("\n"))
                return True
        except Exception:
            continue
    return False


def try_pillow(path, args):
    try:
        from PIL import Image
    except ImportError:
        return False
    img = Image.open(path).convert("L")
    width = args.width or 80
    aspect = img.height / img.width * 0.5  # chars are ~2x tall
    height = max(1, int(width * aspect))
    img = img.resize((width, height))
    px = img.load()
    out = []
    for y in range(height):
        row = []
        for x in range(width):
            v = px[x, y]
            row.append(RAMP[v * (len(RAMP) - 1) // 255])
        out.append("".join(row))
    print("\n".join(out))
    return True


def main():
    p = argparse.ArgumentParser(description="Convert an image to ASCII art")
    p.add_argument("image", help="path to image file")
    p.add_argument("--width", type=int, default=None, help="output width in characters")
    p.add_argument("--color", action="store_true", help="preserve color (chafa only)")
    args = p.parse_args()

    if try_cli(args.image, args):
        return
    if try_pillow(args.image, args):
        return
    print(
        "No image-to-ASCII tool found. Install one of: chafa, jp2a, viu, img2txt "
        "(or Pillow: pip install Pillow).",
        file=sys.stderr,
    )
    sys.exit(1)


if __name__ == "__main__":
    main()

File: skills/ascii-art/scripts/img2ascii.ts
#!/usr/bin/env bun
// img2ascii.ts — convert an image to ASCII/ANSI art in the terminal.
//
// Uses, in order of preference: `chafa`, `jp2a`, `viu`, `img2txt` (libcaca).
// Falls back with an install hint.
//
// Examples:
//   bun scripts/img2ascii.ts logo.png
//   bun scripts/img2ascii.ts photo.jpg --width 80
//   bun scripts/img2ascii.ts photo.jpg --color

import { spawnSync } from "node:child_process";

function tryCli(path: string, width?: number, color = false): boolean {
  const tools: Array<{ cmd: string; base: string[]; widthFlag: boolean }> = [
    {
      cmd: "chafa",
      base: ["--symbols", "block+border+space", "--format", "symbols"],
      widthFlag: false,
    },
    { cmd: "jp2a", base: ["--width"], widthFlag: true },
    { cmd: "viu", base: ["-w"], widthFlag: true },
    { cmd: "img2txt", base: ["-W"], widthFlag: true },
  ];
  for (const t of tools) {
    const argv = [...t.base];
    if (t.widthFlag && width) argv.push(String(width));
    if (t.cmd === "chafa" && width) argv.push("--size", `${width}x`);
    if (t.cmd === "chafa" && color) argv.push("--colors", "full");
    argv.push(path);
    const res = spawnSync(t.cmd, argv, { encoding: "utf8", timeout: 30_000 });
    if (!res.error && res.status === 0 && (res.stdout ?? "").trim()) {
      process.stdout.write(`${(res.stdout ?? "").replace(/\n+$/, "")}\n`);
      return true;
    }
  }
  return false;
}

function main(): void {
  const args = process.argv.slice(2);
  let width: number | undefined;
  let color = false;
  const positional: string[] = [];
  for (let i = 0; i < args.length; i++) {
    const a = args[i]!;
    if (a === "--width") width = Number.parseInt(args[++i] ?? "", 10);
    else if (a === "--color") color = true;
    else if (a === "--help" || a === "-h") {
      console.log("usage: img2ascii.ts [--width N] [--color] IMAGE");
      process.exit(0);
    } else positional.push(a);
  }

  const image = positional[0];
  if (!image) {
    console.error("error: image path required");
    process.exit(2);
  }

  if (tryCli(image, width, color)) return;
  console.error("No image-to-ASCII tool found. Install one of: chafa, jp2a, viu, img2txt.");
  process.exit(1);
}

main();
